crates/bevy_ecs/src/reflect/map_entities.rs

```diff
@@ -12,53 +14,73 @@ use bevy_reflect::FromType;
+    /// by the [`EntityHashMap<Entity>`] might already contain valid entity references, you should use [`map_world_entities`](Self::map_world_entities).
     ///
     /// An example of this: A scene can be loaded with `Parent` components, but then a `Parent` component can be added
-    /// to these entities after they have been loaded. If you reload the scene using [`map_all_entities`](Self::map_all_entities), those `Parent`
+    /// to these entities after they have been loaded. If you reload the scene using [`map_all_world_entities`](Self::map_all_world_entities), those `Parent`
     /// components with already valid entity references could be updated to point at something else entirely.
-    pub fn map_all_entities(&self, world: &mut World, entity_map: &mut EntityHashMap<Entity>) {
-        SceneEntityMapper::world_scope(entity_map, world, self.map_all_entities);
+    #[deprecated = "map_all_world_entities doesn't play well with Observers. Use map_entities instead."]
+    pub fn map_all_world_entities(
+        &self,
+        world: &mut World,
+        entity_map: &mut EntityHashMap<Entity>,
+    ) {
+        SceneEntityMapper::world_scope(entity_map, world, self.map_all_world_entities);
     }
 
     /// A general method for applying [`MapEntities`] behavior to elements in an [`EntityHashMap<Entity>`]. Unlike
-    /// [`map_all_entities`](Self::map_all_entities), this is applied to specific entities, not all values
+    /// [`map_all_world_entities`](Self::map_all_world_entities), this is applied to specific entities, not all values
     /// in the [`EntityHashMap<Entity>`].
     ///
     /// This is useful mostly for when you need to be careful not to update components that already contain valid entity
-    /// values. See [`map_all_entities`](Self::map_all_entities) for more details.
-    pub fn map_entities(
+    /// values. See [`map_all_world_entities`](Self::map_all_world_entities) for more details.
+    #[deprecated = "map_world_entities doesn't play well with Observers. Use map_entities instead."]
+    pub fn map_world_entities(
         &self,
         world: &mut World,
         entity_map: &mut EntityHashMap<Entity>,
         entities: &[Entity],
     ) {
         SceneEntityMapper::world_scope(entity_map, world, |world, mapper| {
-            (self.map_entities)(world, mapper, entities);
+            (self.map_world_entities)(world, mapper, entities);
         });
     }
 }
 
-impl<C: Component + MapEntities> FromType<C> for ReflectMapEntities {
+impl<C: Component + MapEntities + Reflect + FromReflect> FromType<C> for ReflectMapEntities {
     fn from_type() -> Self {
         ReflectMapEntities {
-            map_entities: |world, entity_mapper, entities| {
+            map_entities: |component, entity_mapper| {
+                let mut concrete = C::from_reflect(&*component).unwrap();
+                concrete.map_entities(&mut UnDynEntityMapper(entity_mapper));
+                component.apply(&concrete as _);
```

Comment:
The only possible concern here is if the user has mapped fields marked with `#[reflect(ignore)]`. The original implementation didn't rely on reflection so it was fine, but this is a new "hidden" requirement. We should probably document this somewhere.

Reply:
I feel like that's expected? If a field is invisible to reflection in general, it seems logical that it would be invisible to reflected traits.

Reply:
Hm fair point. And when used with scenes specifically I guess those fields won't even be serialized which means it's less of an issue than I originally thought.